File: extension.ts
//% color="#359eff" weight=20 icon="\uf1b2"
namespace MazeRobot {
    export enum Param{
        //% block="直行速度(默认90)"
        MAX_LINEAR_SPEED        =0x01,
        //% block="转向速度(默认70)"
        MAX_ANGULER_SPEED       =0x02,
        //% block="一格距离(默认175)"
        DIST_PER_CELL           =0x03,
        //% block="贴墙前进时与墙面距离(默认25)"
        WALL_KEEP_DIST          =0x04,
        //% block="探墙阈值(默认60)"
        WALL_DETECT_DIST        =0x05,
        //% block="前后校准阈值(默认140)"
        F_WALL_CORRECT_THRESHOLD=0x06,
        //% block="前后校准目标值(默认25)"
        F_WALL_TARGET_DIST      =0x07,
        //% block="前后校准容差(默认1)"
        F_ERR_MARGIN            =0x08,
        //% block="平行校正容差(默认2)"
        PARALLEL_ERR_MARGIN     =0x09,
        //% block="平行校正次数(默认10)"
        PARALLEL_CORRECT_TIMES  =0x0A,
        //% block="转向补偿(默认0)"
        TURN_COMPENSATION       =0x0B,
    }
    export enum Motor_ID{
        M1,
        M2,
        M3,
        M4
    }
    
    //% shim=MazeRobot::Init
    //% block="初始化机器人 左马达%ML|右马达%MR" 
    export function Init(ML:Motor_ID,MR:Motor_ID) {
        return
    }

    export enum ENABLE{
        //% block="是"
        YES,
        //% block="否"
        NO
    }
    
    //% block="按设定规则运行到X=%x|Y=%y"
    //% shim=MazeRobot::Move_to
    export function Move_to(x:number,y:number){
        return
    }
    
    //% block="按照记录的路径前进"
    //% shim=MazeRobot::Go_Path
    export function Go_Path(){
        return
    }
    
    //% block="贴墙前进%n格长度"
    //% shim=MazeRobot::Go_Distance
    export function Go_Distance(n:number){
        return
    }

    export enum TURNS{
        //% block="右转90°"
        RIGHT_90=1,
        //% block="左转90°"
        LEFT_90=2,
        //% block="掉头180°"
        TURN_180=3,
        //% block="直行"
        GO_STRAIGHT=0
    }

    //% block="原地%turn"
    //% shim=MazeRobot::Turn
    export function Turn(turn:TURNS){

    }

    export enum DIR{
        //% block="上"
        UP,
        //% block="下"
        DOWN=2,
        //% block="左"
        LEFT=3,
        //% block="右"
        RIGHT=1
    }

    //% block="设定当前位置为X=%x|Y=%y|面朝方向%dir"
    //% shim=MazeRobot::Set_Pos
    export function Set_Pos(x:number,y:number,dir:DIR){
        return
    }
    
    //% block="按照字符串前进%path"
    export function Go_Path_Custom(path:string){
        let cmd=pins.createBuffer(2);
        cmd[0]=7;
        cmd[1]=path.length;
        pins.i2cWriteBuffer(0x70,cmd,false);
        let buf=pins.createBuffer(path.length);
        for(let i=0;i<path.length;i++){
            buf[i]=path.charCodeAt(i);
            // console.log(buf[i].toString());
        }
        basic.pause(10);
        pins.i2cWriteBuffer(0x70,buf,false);
        basic.pause(100);
        while(1){
            pins.i2cWriteNumber(0x70,0,NumberFormat.UInt16LE,false);
            basic.pause(10);
            if(pins.i2cReadNumber(0x70,NumberFormat.UInt8LE,false)==1)
                break;
            basic.pause(500);
        }
        return
    }
    
    //% block="平行校准"
    //% shim=MazeRobot::Paraller_Correct
    export function Paraller_Correct(){
        return
    }
    
    //% block="前后校准"
    //% shim=MazeRobot::FB_Correct
    export function FB_Correct(){
        return
    }

    export enum Regular{
        //% block="靠左墙"
        FOLLOW_LEFT,
        //% block="靠右墙"
        FOLLOW_RIGHT,
        //% block="自定义"
        CUSTOM
    }

    //% block="设定选方向规则为%regular"
    //% shim=MazeRobot::Set_Regular
    export function Set_Regular(regular:Regular){
        return
    }

    //% block="发送传感器数据"
    //% shim=MazeRobot::Get_Sensor_Data
    export function Get_Sensor_Data(){
        return
    }

    export enum Sensor_ID{
        //% block="左前"
        LEFT_FRONT,
        //% block="左后"
        LEFT_REAR,
        //% block="右前"
        RIGHT_FRONT,
        //% block="右后"
        RIGHT_REAR,
        //% block="正前"
        MID_FRONT,
    }
    
    //% shim=MazeRobot::Set_Param
    //% block="设置控制参数 %param | %value" 
    //% value.defl=175
    export function Set_Param(param:Param,value:number) {
        return
    }   
    
}

File: extension.cpp
#include "pxt.h"
#include "i2c_hw_interface.h"
using namespace pxt;

#define 	DETECT 							0x01	//三个参数，目标(x,y)和是否记录并简化路径
#define 	SEND_PATH						0x02	//发送记录的路径
#define		RUSH							0x03	//若记录并简化路径则可按照简化后的路径RUSH
#define 	DISTANCE_MOVE   				0x04	//前进一段距离
#define 	TURN   							0x05	//turn('S/L/R/B') 注：会记录坐标变化
#define 	SET_POSITION   					0x06	//设定当前坐标与朝向
#define 	STRING_MOVE   					0x07	//接收一个字符串，按照字符串前进
#define 	PARALLEL_CORRECTION   	        0x08	
#define 	FRONT_CORRECTION   			    0x09	
#define 	SET_TURN_RULE   				0x0A	//设置LR，LR=0左迷宫，LR=1右迷宫
#define 	SET_MOTOR						0x0B	//设置电机序号
#define 	SEND_SENSOR_VALUE   		    0x0C	//发送传感器数值0~5->右后/右前/正前/左前/左后
#define 	SENSOR_RESULT  					0x0D	//发送传感器结果（是否有墙）

#define 	SET_STRAIGHT_SPEED  		    0x0E
#define 	SET_TURN_SPEED  				0x0F
#define 	SET_DIST_PER_CELL  			    0x10
#define 	SET_LR_TARGET_DIST  		    0x11
#define		SET_WALL_EXIST_DIST 		    0x12
#define		SET_F_CORRECT_DIST			    0x13
#define		SET_F_TARGET_DIST				0x14
#define		SET_F_MARGIN					0x15
#define		SET_PARALLEL_MARGIN			    0x16
#define		SET_PARALLEL_CORRECT_TIMES	    0x17
#define		SET_TURN_COMPENSATION		    0x18
namespace MazeRobot {
    CubeI2C* i2c=new CubeI2C(&uBit.i2c,0x70);
    uint8_t Left_wheel=255;
    uint8_t Right_wheel=255;
    uint8_t data[16]={0};
    //%
    void wait_for_finish(){
        uint8_t is_finish=0;
        while(1){
            if(i2c->I2CRead(0x00,&is_finish)!=0)
                continue;
            else
                if(is_finish)
                    break;
            wait_ms(500);
        }
    }

    //%
    void Init(int ML,int MR) {
        uBit.io.P16.setDigitalValue(0);
        wait_us(100);
        uBit.io.P16.setDigitalValue(1);
        wait_ms(100);
        i2c->I2CWrite2Bytes(SET_MOTOR,0);
        wait_ms(10);
        i2c->I2CWrite2Bytes(ML+1,MR+1);
        wait_ms(10);
    }

    //%
    void Go_Path() {
        i2c->I2CWrite2Bytes(RUSH,0);
        wait_ms(500);
        wait_for_finish();
    }

    //%
    void Go_Distance(int n) {
        i2c->I2CWrite2Bytes(DISTANCE_MOVE,(uint8_t)n);
        wait_ms(500);
        wait_for_finish();
    }

    //%
    void Turn(int dir) {
        i2c->I2CWrite2Bytes(TURN,(uint8_t)dir);
        wait_ms(500);
        wait_for_finish();
    }

    //% 
    void Set_Pos(int x,int y,int dir) {
        i2c->I2CWrite2Bytes(SET_POSITION,0);
        uint8_t cmd[3];
        cmd[0]=x;
        cmd[1]=y;
        cmd[2]=dir;
        wait_ms(10);
        i2c->I2CWriteBuffer(cmd,3);
        wait_ms(10);
    }

    //%
    void Paraller_Correct() {
        i2c->I2CWrite2Bytes(PARALLEL_CORRECTION,0);
        wait_ms(500);
        wait_for_finish();
    }
    //%
    void FB_Correct() {
        i2c->I2CWrite2Bytes(FRONT_CORRECTION,0);
        wait_ms(500);
        wait_for_finish();
    }

    //%
    void Set_Regular(int regular) {
        i2c->I2CWrite2Bytes(SET_TURN_RULE,regular);
        wait_ms(10);
    }

    //%
    void Set_Param(int param_id, int value) {
        i2c->I2CWrite2Bytes(param_id+0x0e,0);
        wait_ms(10);
        i2c->I2CWrite2Bytes((uint8_t)(value&0x00ff),(uint8_t)(value>>8));
        wait_ms(10);
    }
    
    //%
    void Move_to(int x,int y) {
        i2c->I2CWrite2Bytes(DETECT,1);
        wait_ms(10);
        i2c->I2CWrite2Bytes(x,y);
        wait_ms(1000);
        wait_for_finish();
    }
    
    //%
    void Get_Sensor_Data(){
        i2c->I2CWrite2Bytes(SEND_SENSOR_VALUE,1);
    }

}